### PR TITLE
[WIP] os/env.c: Fix bug where empty $HOME breaks init_homedir().

### PR DESCRIPTION
**Pull request by tomeaton17**
When HOME environment variable is not set in UNIX environment,
init_homedir() fails. This provides an else clause using uv_os_homedir()
to get the current users homedirectoy when HOME is not set.

This is my second attempt. I have tried to implement everything mentioned by @jamessan in my previous PR #8641.

The setting of windows environment variables is untested. 

Tests are failing, mainly because `test/helpers.lua:190: attempt to index a nil value.` I assume this is because I am not setting something?

### CODE REVIEW

**Comment by janlazo** [comment]
Why did you delete the Windows code? Did you consult the docs from libuv and Microsoft? Also, why are you setting `$HOME`?

**Comment by tomeaton17** [comment]
I was following the comments by @jamessan on #8641 The windows part is not complete yet (as need to test on Windows machine) but according to the libuv docs this should work.

> Gets the current user’s home directory. On Windows, uv_os_homedir() first checks the USERPROFILE environment variable using GetEnvironmentVariableW(). If USERPROFILE is not set, GetUserProfileDirectoryW() is called.

As for setting $HOME, unless I have misunderstood this was suggested in the comments on #8641

**Comment by janlazo** [comment]
The code you removed uses `HOMEDRIVE`, `HOMEPATH`, not `USERPROFILE`. #8485 checks `USERPROFILE` but it fails with the new-style oldtests included in the patches.

**Comment by jamessan** [comment]
From what searching I've done, it sounds to me like Vim's use of `HOMEDRIVE` and `HOMEPATH` aren't recommended.  `USERPROFILE` (with a fallback to the API call if the variable isn't present) should be the preferred mechanism, which is exactly what libuv does.

My only Windows systems are work computers and they all have networked drives for `HOMEDRIVE`/`HOMEPATH`, which I specifically override by adding an explicit `HOME` (same value as `USERPROFILE`) to avoid using the networked drive.  So, sample size of one, I would generally prefer what libuv does over what Vim does.

**Comment by tomeaton17** [comment]
Okay. As for the code, what needs to be done?

**Comment by ZviRackover** [comment]
Doesn't nvim support Windows XP or later? In that case maybe it's ok to diverge from vim which i assume supports much older versions of Windows?

**Comment by ZviRackover** [comment]
Would the following implementation of init_home() work?
```
/// Attempts to set 'homedir'.
void init_homedir(void)
{
  // In case we are called a second time.
  xfree(homedir);
  homedir = NULL;

  size_t buf_len = IOSIZE;
  int status;
  if (!(status = uv_os_homedir((char *)IObuff, &buf_len))) {
    // This is the likely path where the call succeeds.
    homedir = xstrdup((const char *)IObuff);
    return;
  }
  if (UV_ENOBUFS == status) {
    // Need a larger buffer for the result to fit. Allocate on the heap.
    char *homedir_buf = xmalloc(buf_len);
    if (NULL == homedir_buf) {
      // Out of memory
      return;
    }
    if ((status = uv_os_homedir((char *)homedir_buf, &buf_len))) {
      xfree(homedir_buf);
      return;
    }
    // Success
    homedir = homedir_buf;
    homedir_buf = NULL;
  }
}
```

**Comment by janlazo** [comment]
My position on this matter is which subdirectory is `%LOCALAPPDATA%` in. Is it  `%HOMEDRIVE%%HOMEPATH%` or `%USERPROFILE%`? If the former, then keep the Vim code. If the latter, then it's acceptable to diverge from Vim and use `%USERPROFILE%` and libuv APIs. Idea is to be consistent with the paths given by `stdpath()`.

**Comment by tomeaton17** [comment]
According to the [Microsoft documentation on folder variables](https://www.microsoft.com/en-us/wdsi/help/folder-variables), `%LOCALAPPDATA%` and `%HOMEDRIVE%%HOMEPATH%` are equal.

**Comment by janlazo** [comment]
@tomeaton17  You meant `%USERPROFILE%`?

**Comment by tomeaton17** [comment]
Apologies, I have edited my previous comment.

**Comment by tomeaton17** [comment]
@janlazo `%LOCALAPPDATA%` seems to point to the local drive by default but can be changed by user to point to a network drive (why they would want to do this I don't know). `%LOCALAPPDATA%` is designed to be used for configuration so I think it would be a good choice.

File: src/nvim/os/env.c [hunk 1]
```diff
@@ -187,40 +187,28 @@ void init_homedir(void)
...
+  if (var) {
+    is_var = 1; 
+  }
+  size_t max_path_size = 2000;
+
+  if (!uv_os_homedir((char *)IObuff, &max_path_size)) {
+    var = (char *)IObuff;
+    if (!is_var) {
+#ifdef UNIX    
+      os_setenv("HOME", var, 1);
```
Comment (janlazo) [inline src/nvim/os/env.c]: What's this for?

File: src/nvim/os/env.c [hunk 2]
```diff
@@ -187,40 +187,28 @@ void init_homedir(void)
...
-    if (homepath == NULL) {
-        homepath = "\\";
-    }
-    if (homedrive != NULL && strlen(homedrive) + strlen(homepath) < MAXPATHL) {
-      snprintf(os_buf, MAXPATHL, "%s%s", homedrive, homepath);
-      if (os_buf[0] != NUL) {
-        var = os_buf;
-        vim_setenv("HOME", os_buf);
-      }
+      vim_setenv("HOME", var);
```
Comment (janlazo) [inline src/nvim/os/env.c]: Remove this to match the changes in 8.0.1012.

File: src/nvim/os/env.c [hunk 3]
```diff
@@ -187,40 +187,28 @@ void init_homedir(void)
...
+#endif
     }
   }
-#endif
-
+  else {
+    printf("failed");
```
Comment (janlazo) [inline src/nvim/os/env.c]: Is this for debugging only?

File: src/nvim/os/env.c [hunk 4]
```diff
@@ -187,40 +187,28 @@ void init_homedir(void)
...
   const char *var = os_getenv("HOME");
 
+  bool is_var = 0;
+
+  if (var) {
+    is_var = 1; 
+  }
+  size_t max_path_size = 2000;
+
+  if (!uv_os_homedir((char *)IObuff, &max_path_size)) {
```
Comment (justinmk) [inline src/nvim/os/env.c]: This should be:

```c
int rv = uv_os_homedir((char *)IObuff, &max_path_size);
if (rv != 0) {
  emsgf("E00: %d: %s", rv, os_strerror(rv));
  ELOG("E00: %d: %s", rv, os_strerror(rv));
}
```

**Comment by justinmk** [comment]
https://github.com/neovim/neovim/pull/8643#issuecomment-400088574 makes sense to me, but should we leave the old fallback as a last resort (if `uv_os_homedir` fails)?

**Comment by justinmk** [comment]
Thanks for getting this started. Continued at #11140, see https://github.com/neovim/neovim/pull/11140#issuecomment-544808617

**Pull request closed without merging**